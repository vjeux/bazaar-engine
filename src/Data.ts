import pako from "pako";
import type { Cards } from "./types/cardTypes.ts";
import type { EncounterDays } from "./types/encounterTypes.ts";

function genDataFromLocalStorage(
  localStorageKey: string,
  importCall: Promise<any>
) {
  return new Promise((success) => {
    if (typeof localStorage === "undefined") {
      importCall.then((module) => {
        success(module.default);
      });
    } else {
      const storage = localStorage.getItem(localStorageKey);
      if (storage == null) {
        importCall.then((module) => {
          const data = module.default;
          const compressed = JSON.stringify([
            ...pako.deflate(JSON.stringify(data))
          ]);
          try {
            localStorage.setItem(localStorageKey, compressed);
          } catch (e) {
            // Ignore quota errors
          }
          success(data);
        });
      } else {
        const compressed = new Uint8Array(JSON.parse(storage));
        const data = JSON.parse(pako.inflate(compressed, { to: "string" }));
        success(data);
      }
    }
  });
}

export async function genCardsAndEncounters(): Promise<{
  Cards: Cards;
  Encounters: EncounterDays;
}> {
  const [Cards, Encounters] = await Promise.all([
    // https://cdn.playthebazaar.com/bazaardesigndataprod/cards.json
    genDataFromLocalStorage("Cards", import("./json/cards.json")),
    // https://www.howbazaar.gg/api/monsterEncounterDays
    genDataFromLocalStorage(
      "Encounters",
      import("./json/monsterEncounterDays.json")
    )
  ]);
  return { Cards: Cards as Cards, Encounters: Encounters as EncounterDays };
}
