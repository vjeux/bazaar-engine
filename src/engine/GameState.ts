import type { BoardCard, BoardSkill, GameState, Player } from "./Engine.ts";

// Import JSON types generated by quicktype
import type { Card, Cards, Enchantments } from "../types/cardTypes.ts";
import type { Tier } from "../types/shared.ts";
import type { EncounterDays, Group } from "../types/encounterTypes.ts";
import { z } from "zod";
/* Bugs

// Day 1 - Viper
// Toxic Fang's poison is 10% of 5 = 0.5. In combat it does 1 damage but outside it shows 0.
// Not sure how to fix this.
// https://www.youtube.com/watch?v=nPYXguaNr8I

// Day 4 - Flame Juggler
// The skill Burst of Flame seems to trigger when the oppononent reaches 1/3 of life rather than 1/2.

// Need to fix crit trigger

*/

function _createBoardCardFromCard(
  card: Card,
  tier: Tier,
  enchantment: keyof Enchantments | null = null,
): BoardCard {
  let attributes: Partial<BoardCard> = {
    Abilities: card.Abilities,
    Auras: card.Auras,
    Localization: {
      Tooltips: card.Localization.Tooltips,
      Title: {
        Text: card.Localization.Title.Text,
      },
    },
  };

  // If the provided tier is not available, default to the first tier available.
  if (!card.Tiers || !(tier in card.Tiers)) {
    // Get tier key
    const tierKeys = Object.keys(card.Tiers || {});
    tier = tierKeys[0] as Tier;
  }

  // Iterate over the tiers in order and merge their attributes until the desired tier is reached.
  for (const [tn, tierValues] of Object.entries(card.Tiers || {})) {
    attributes = {
      ...attributes,
      ...(tierValues?.Attributes ?? {}),
      AbilityIds: tierValues?.AbilityIds || attributes.AbilityIds,
      AuraIds: tierValues?.AuraIds || attributes.AuraIds,
      TooltipIds: tierValues?.TooltipIds || attributes.TooltipIds,
    };

    if (tn === tier) {
      break;
    }
  }

  if (enchantment) {
    if (!card.Enchantments) {
      throw new Error(
        `No enchantments available for card ${card.Localization.Title.Text}`,
      );
    }
    const enchant = card.Enchantments[enchantment];
    if (!enchant) {
      throw new Error(
        `Enchantment ${enchantment} not found for card ${card.Localization.Title.Text}`,
      );
    }
    attributes = { ...attributes, ...enchant.Attributes };
    if (enchant.HasAbilities) {
      attributes.Abilities = {
        ...attributes.Abilities,
        ...enchant.Abilities,
      };
      attributes.AbilityIds = [
        ...attributes.AbilityIds,
        ...Object.keys(enchant.Abilities),
      ];
    }
    if (enchant.HasAuras) {
      attributes.Auras = { ...attributes.Auras, ...enchant.Auras };
      attributes.AuraIds = [
        ...attributes.AuraIds,
        ...Object.keys(enchant.Auras),
      ];
    }
    attributes.Tags = [...(attributes.Tags ?? []), ...enchant.Tags];
    attributes.HiddenTags = [
      ...(attributes.HiddenTags ?? []),
      ...enchant.HiddenTags,
    ];
    attributes.Localization.Tooltips = [
      ...attributes.Localization.Tooltips,
      ...enchant.Localization.Tooltips,
    ];
    attributes.TooltipIds = [
      ...attributes.TooltipIds,
      ...enchant.Localization.Tooltips.map((tooltip) =>
        attributes.Localization.Tooltips.indexOf(tooltip),
      ),
    ];
    attributes.Localization.Title.Text = `${enchantment} ${attributes.Localization.Title.Text}`;
  }

  const result = {
    card: card,
    tick: 0,
    Slow: 0,
    Freeze: 0,
    Haste: 0,
    CritChance: 0,
    DamageCrit: 0,
    tier: tier,
    Enchantment: enchantment,
    isDisabled: false,
    ...attributes,
  };

  return result;
}

function getBoardCard(
  Cards: Cards,
  name: string,
  tier: Tier | null = null,
  enchantment: keyof Enchantments | null = null,
): BoardCard {
  const CardsValues = Cards["0.1.9"];
  const card = CardsValues.find((c) => c.Localization?.Title?.Text === name);
  if (!card) {
    throw new Error(`Card ${name} not found`);
  }
  return _createBoardCardFromCard(card, tier ?? card.StartingTier, enchantment);
}

function getBoardCardFromId(
  Cards: Cards,
  cardId: string,
  tier: Tier,
  enchantment: keyof Enchantments | null = null,
): BoardCard {
  const card = Cards["0.1.9"].find((card) => card.Id === cardId);
  if (!card) {
    throw new Error(`Card from id ${cardId} not found`);
  }
  return _createBoardCardFromCard(card, tier, enchantment);
}

function getBoardSkill(
  Cards: Cards,
  name: string,
  tier: Tier | null = null,
  modifiers: any = {},
): BoardSkill {
  const CardsValues = Cards["0.1.9"];
  const card = CardsValues.find(
    (card) => card.Localization.Title.Text === name,
  );
  if (!card) {
    throw new Error(`Card for Skill ${name} not found`);
  }
  let attributes: Partial<BoardSkill> = {
    Abilities: card.Abilities,
    Auras: card.Auras,
    Localization: {
      Tooltips: card.Localization.Tooltips,
      Title: {
        Text: card.Localization.Title.Text,
      },
    },
  };
  if (!card.Tiers || !card.Tiers[tier ?? card.StartingTier]) {
    throw new Error(
      name +
        " doesn't have tier " +
        tier +
        ", the first one is " +
        Object.keys(card.Tiers || {})[0],
    );
  }
  const tierNames = Object.keys(card.Tiers) as Tier[];
  for (let i = 0; i < tierNames.length; ++i) {
    const tierName = tierNames[i];
    const tierValues = card.Tiers[tierName];
    attributes = {
      ...attributes,
      ...tierValues?.Attributes,
      AbilityIds: tierValues?.AbilityIds,
      AuraIds: tierValues?.AuraIds,
      TooltipIds: tierValues?.TooltipIds,
    };
    if (tierName === (tier ?? card.StartingTier)) {
      break;
    }
  }
  const result = {
    card,
    ...attributes,
    tier: tier ?? card.StartingTier,
  };
  return result;
}

function getBoardPlayer(
  stats: Omit<Partial<Player>, "board"> & { HealthMax: number },
  boardCards: BoardCard[],
  boardSkills: BoardSkill[],
): Player {
  return {
    HealthMax: stats.HealthMax,
    Health: stats.HealthMax,
    HealthRegen: stats.HealthRegen ?? 0,
    Shield: 0,
    Burn: 0,
    Poison: 0,
    Gold: 0,
    Income: 0,
    Hero: "<Hero Name>",
    board: [...boardCards, ...boardSkills],
  };
}

function getBoardPlayerFromMonsterCard(Cards: Cards, monsterCard: Group) {
  return getBoardPlayer(
    { HealthMax: monsterCard.health },
    monsterCard.items.map((item) =>
      getBoardCardFromId(
        Cards,
        item.card.id,
        item.tierType,
        item.enchantmentType,
      ),
    ),
    monsterCard.skills.map((item) =>
      getBoardCardFromId(Cards, item.card.id, item.tierType, null),
    ),
  );
}

function getBoardMonster(
  Cards: Cards,
  Encounters: EncounterDays,
  name: string,
) {
  for (let i = 0; i < Encounters.data.length; ++i) {
    const day = Encounters.data[i];
    for (let j = 0; j < day.groups.length; ++j) {
      const group = day.groups[j];
      for (let k = 0; k < group.length; ++k) {
        const monsterCard = group[k];
        if (monsterCard.cardName === name) {
          return getBoardPlayerFromMonsterCard(Cards, monsterCard);
        }
      }
    }
  }
  throw new Error(`Can't find a monster with name ${name}`);
}

function sfc32(a: number, b: number, c: number, d: number) {
  return function () {
    a |= 0;
    b |= 0;
    c |= 0;
    d |= 0;
    let t = (((a + b) | 0) + d) | 0;
    d = (d + 1) | 0;
    a = b ^ (b >>> 9);
    b = (c + (c << 3)) | 0;
    c = (c << 21) | (c >>> 11);
    c = (c + t) | 0;
    return (t >>> 0) / 4294967296;
  };
}

// Define Zod schema for MonsterConfig
export const MonsterConfigSchema = z.object({
  type: z.literal("monster"),
  name: z.string(),
});

// Infer the TypeScript type from the Zod schema
export type MonsterConfig = z.infer<typeof MonsterConfigSchema>;

export interface PlayerCardConfig {
  name: string;
  tier?: Tier;
  enchantment?: keyof Enchantments | null;
}

export interface PlayerSkillConfig {
  name: string;
  tier?: Tier;
}

export interface PlayerConfig {
  type: "player";
  health?: number;
  healthRegen?: number;
  cards?: PlayerCardConfig[];
  skills?: PlayerSkillConfig[];
}

export function getInitialGameState(
  Cards: Cards,
  Encounters: EncounterDays,
  config: (MonsterConfig | PlayerConfig)[],
): GameState {
  return {
    tick: 0,
    isPlaying: true,
    players: config.map((player) => {
      if (player.type === "monster") {
        if (!player.name) throw new Error("Monster name is required");
        return getBoardMonster(Cards, Encounters, player.name);
      } else {
        return getBoardPlayer(
          {
            HealthMax: player.health ?? 3500,
            HealthRegen: player.healthRegen ?? 0,
          },
          (player.cards ?? []).map((c) =>
            getBoardCard(Cards, c.name, c.tier, c.enchantment),
          ),
          (player.skills ?? []).map((s) =>
            getBoardSkill(Cards, s.name, s.tier),
          ),
        );
      }
    }),
    multicast: [],
    getRand: sfc32(0, 10000, 10000000, 100000000000),
  };
}

export function getFlattenedEncounters(Encounters: EncounterDays): Array<{
  name: string;
  card: { cardName: string; cardId: string };
  day: number | string;
}> {
  return Encounters.data
    .map((data) => {
      return data.groups
        .map((day) => {
          return day.map((group) => ({
            name: group.cardName,
            card: { cardName: group.cardName, cardId: group.cardId },
            day: data.day,
          }));
        })
        .flat();
    })
    .flat();
}
