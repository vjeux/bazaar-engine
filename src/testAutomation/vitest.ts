import { createVitest, startVitest } from "vitest/node";
import apiItems from "../json/items.json" with { type: "json" };
import { Item } from "../types/apiItems.ts";

const items: Item[] = apiItems.data as Item[];

const vitest = await createVitest(
  "test",
  {
    watch: false
  },
  {}, // override Vite config
  {} // custom Vitest options
);

try {
  const result = await vitest.start(["peggy"]);

  // Collect all failed tasks
  const parseToCardTasks = result.testModules[0].task.tasks[0].tasks;

  // Filter for failing tasks
  const failingTasks = parseToCardTasks.filter(
    (task) => task.result.state === "fail"
  );

  // Take the first failing task
  const firstFailingTask = failingTasks[2];

  // Get the different parts needed for llm to generate solution
  const stack = firstFailingTask.result.errors[0].stack;
  const errorMessage = firstFailingTask.result.errors[0].message;
  const expected = firstFailingTask.result.errors[0].expected;
  const actual = firstFailingTask.result.errors[0].actual;

  // Get peggy code, read file
  const peggyCode = await Deno.readTextFile("src/tooltipParser.pegjs");

  // Read peggy documentation from file
  const peggyDoc = await Deno.readTextFile(
    "src/testAutomation/documentation.html"
  );

  const systemPrompt =
    `` +
    `You are an experienced software engineer\n` +
    `You are an expert in parsing and generating code\n` +
    `You are an expert in pegjs and peggy\n` +
    `You are building a program that takes tooltips of a card game and your goal is to generate the internal json representation of the card.\n` +
    `You are going to be given a tooltip and the expected card json.\n` +
    `Your goal is to write a generalizable peggy/pegjs parser that can parse the tooltip and generate the json.\n` +
    `The parser should be able to handle many different tooltips, and should be modular and reusable.\n` +
    `You will also be given the error message and stack trace of the test that failed.\n` +
    `When responding, you should only respond with the whole code of the parser.`;

  // Get the card name from task name
  const cardName = firstFailingTask.name.split("-")[0];

  console.log("Card name:", cardName);

  // Get the tooltip by filtering items
  const tooltips = JSON.stringify(
    items.find((item) => item.name.toLowerCase() === cardName.toLowerCase())
      ?.unifiedTooltips
  );

  console.log("Tooltip:", tooltips);

  // Create prompt by combining everything
  const prompt =
    `<systemPrompt>\n${systemPrompt}\n</systemPrompt>\n\n` +
    `<tooltips>\n${tooltips}\n</tooltips>\n\n` +
    `<cardName>\n${cardName}\n</cardName>\n\n` +
    `<peggyCode>\n${peggyCode}\n</peggyCode>\n\n` +
    `<peggyDoc>\n${peggyDoc}\n</peggyDoc>\n\n` +
    `<errorMessage>\n${errorMessage}\n</errorMessage>\n\n` +
    `<stackTrace>\n${stack}\n</stackTrace>\n\n` +
    `<expected>\n${expected}\n</expected>\n\n` +
    `<actual>\n${actual}\n</actual>\n\n`;

  // Store prompt in file
  await Deno.writeTextFile("src/testAutomation/prompt.txt", prompt);

  // write test result to file
  // const cache: any[] = [];
  // await Deno.writeTextFile(
  //   "test-result.json",
  //   JSON.stringify(result, (key, value) => {
  //     if (typeof value === "object" && value !== null) {
  //       // Duplicate reference found, discard key
  //       if (cache.includes(value)) return;

  //       // Store value in our collection
  //       cache.push(value);
  //     }
  //     return value;
  //   })
  // );

  console.log("Prompt written to prompt.txt");
} catch (err) {
  console.log("Error starting Vitest:", err);
} finally {
  await vitest.close();
}
