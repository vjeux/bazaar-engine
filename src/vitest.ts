import { createVitest, startVitest } from "vitest/node";

const vitest = await createVitest(
  "test",
  { watch: false }, // override test config
  {}, // override Vite config
  {} // custom Vitest options
);

try {
  const result = await vitest.start(["peggy"]);

  console.log("Vitest result:", result);
} catch (err) {
  console.log("Error starting Vitest:", err);
} finally {
  await vitest.close();
}
