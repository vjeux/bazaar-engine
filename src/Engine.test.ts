import { genCardsAndEncounters } from "./Data";
import { run } from "./Engine";
import { getFlattenedEncounters, getInitialGameState } from "./GameState";
import { expect, describe, test } from "vitest";

function getStateDiff(
  prev: any,
  curr: any,
  path: string[] = []
): Record<string, any> {
  const diff: Record<string, any> = {};

  if (typeof prev !== "object" || typeof curr !== "object") {
    if (prev !== curr) {
      diff[path.join(".")] = curr;
    }
    return diff;
  }

  if (prev === null || curr === null) {
    if (prev !== curr) {
      diff[path.join(".")] = curr;
    }
    return diff;
  }

  const prevKeys = Object.keys(prev);
  const currKeys = Object.keys(curr);
  const allKeys = new Set([...prevKeys, ...currKeys]);
  allKeys.delete("card");
  allKeys.delete("tick");

  for (const key of allKeys) {
    if (!(key in prev)) {
      diff[path.concat(key).join(".")] = curr[key];
    } else if (!(key in curr)) {
      diff[path.concat(key).join(".")] = undefined;
    } else {
      const valueDiff = getStateDiff(prev[key], curr[key], path.concat(key));
      Object.assign(diff, valueDiff);
    }
  }

  return diff;
}

describe("Engine encounters", async () => {
  const { Cards, Encounters } = await genCardsAndEncounters();

  getFlattenedEncounters(Encounters).forEach((encounter) => {
    test(`runs encounter "Day ${encounter.day} - ${encounter.name}" against empty board`, () => {
      console.log(encounter.day, encounter.name);
      const gameState = getInitialGameState(Cards, Encounters, [
        { type: "monster", name: encounter.name },
        { type: "player", health: 1000 }
      ]);

      let steps = [];
      try {
        steps = run(gameState);
      } catch (e) {
        steps = [{ error: "null" }, { error: (e as Error).message }];
      }

      const diffs: any[] = [];
      for (let i = 1; i < steps.length; i++) {
        const diff = getStateDiff(steps[i - 1], steps[i]);
        if (Object.keys(diff).length > 0) {
          diff.step = i;
          diffs.push(diff);
        }
      }

      expect(diffs).toMatchSnapshot();
    });
  });
});
