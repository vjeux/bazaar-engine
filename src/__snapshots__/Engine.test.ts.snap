// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Engine encounters > runs encounter "Day 1 - Banannabal" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 320,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 1 - Fanged Inglet" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 319,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 1 - Haunted Kimono" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 321,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 1 - Kyver Drone" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 319,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 1 - Pyro" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 319,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 1 - Viper" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 318,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 2 - Boarrior" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 327,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 2 - Coconut Crab" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 329,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 2 - Covetous Thief" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 329,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 2 - Giant Mosquito" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 325,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 2 - Rogue Scrapper" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 332,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 3 - Dabbling Apprentice" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 328,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 3 - Eccentric Etherwright" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 250,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 3 - Frost Street Challenger" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 333,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 3 - Rogue Scrapper" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 332,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 3 - Scout Trooper" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 325,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 3 - Tempest Flamedancer" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 327,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 4 - Bloodreef Raider" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 338,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 4 - Boilerroom Brawler" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 340,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 4 - Deadly Crooner" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 313,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 4 - Flame Juggler" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 316,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 4 - Hydrodude" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 337,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 4 - Outlands Dervish" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 323,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 4 - Retiree" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 328,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 4 - Techno Virus" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 322,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 5 - Gorgon Noble" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 221,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 5 - Hakurvian Rocket Trooper" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 335,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 5 - Infernal Envoy" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 196,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 5 - Mod Squad" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 193,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 5 - Preening Duelist" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 120,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 5 - Sabretooth" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 320,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 5 - Trashtown Mayor" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 333,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 6 - Dire Inglet" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 326,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 6 - Dire Mosquito" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 339,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 6 - Enclave Weeper" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 318,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 6 - Foreman" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 268,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 6 - Infernal" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 106,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 6 - Lich" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 161,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 6 - Loan Shark" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 200,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 6 - Trash Golem" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 328,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 6 - Zookeeper" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 301,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 7 - Chilly Charles" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 341,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 7 - Cosmic Roc" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 332,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 7 - Joyful Jack" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 324,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 7 - Radiant Corsair" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 186,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 7 - Sergeant Suds" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 246,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 7 - Shock Trooper" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 86,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 7 - Thug" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 161,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 7 - Treasure Turtle" against empty board 1`] = `
[
  {
    "error": "Unhandled action type: TActionCardForceUse",
    "step": 1,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 7 - Viper Tyrant" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 211,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 8 - Bouncertron" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 120,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 8 - Burninator Bot" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 250,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 8 - Car Conductor" against empty board 1`] = `
[
  {
    "error": "Unhandled action type: TActionCardForceUse",
    "step": 1,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 8 - Cosmic Roc" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 332,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 8 - Dr. Vortex" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 290,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 8 - Infernal Frigate" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 111,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 8 - Joyful Jack" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 324,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 8 - Oasis Guardian" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 231,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 8 - Wandering Shoal" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 201,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 9 - Bloodreef Captain" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 323,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 9 - Death Knight Reaper" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 200,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 9 - Elite Duelist" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 130,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 9 - Enclave Revenant" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 180,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 9 - Ferros Khan" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 77,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 9 - Oasis Guardian" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 231,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 9 - Roaming Isle" against empty board 1`] = `
[
  {
    "error": "Unhandled action type: TActionCardForceUse",
    "step": 1,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 9 - Weapons Platform" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 105,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Awakened District" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 100,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Boss Harrow" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 123,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Frost Street Champion" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 335,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Hulking Experiment" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 146,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Lord Arken" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 68,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Lord of the Wastes" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 250,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Master Alchemist" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 301,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Property Baron" against empty board 1`] = `
[
  {
    "error": "Unhandled action type: TActionCardForceUse",
    "step": 1,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Trash Titan" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 121,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Veteran Octopus" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 150,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Void Colossus" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 176,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Void Golem" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 320,
  },
]
`;

exports[`Engine encounters > runs encounter "Day 10 - Volkas Enforcer" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 146,
  },
]
`;

exports[`Engine encounters > runs encounter "Day event - Bounty Hunter" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 200,
  },
]
`;

exports[`Engine encounters > runs encounter "Day event - Mimic" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 80,
  },
]
`;

exports[`Engine encounters > runs encounter "Day event - Mr. Moo" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 140,
  },
]
`;

exports[`Engine encounters > runs encounter "Day event - Sparring Partner" against empty board 1`] = `
[
  {
    "isPlaying": false,
    "step": 341,
  },
]
`;
